### ISSUE TITLE
Why is prelude not installed via package.el?

### ISSUE THREAD

**Issue by Fuco1** [bbatsov/prelude#1064]
There seem to be some problems with the installer: https://github.com/magnars/dash.el/issues/208

This seems like the wrong way to manage packages in emacs, plus someone even keeps making apt packages :O

If you can give any insight into the issue, I welcome the help, but I feel like this is not something I should support :/

**Comment by AlainODea** [comment]
@Fuco1 the APT package is not part of Prelude and isn't the recommended method of installation.

I get where you are coming from on package.el and I'll explain why I think Prelude is close to what I think you intend for good Emacs delivery. Emacs Prelude is not an Emacs package, it's more of an Emacs distribution (it's a shared ~/.emacs.d/). It wouldn't be appropriate for an Emacs package to modify the global configuration so package.el is not an appropriate installation approach for Prelude itself.

However, Prelude does use package.el internally (through Cask) to manage it's own dependencies, including **dash**. My guess is that the **dash** require bug @zobbo reported is in there somewhere.

Here's my understanding of the high-level path Prelude takes to install **dash**:
1. [**dash** is added the **prelude-packages** list in **prelude-packages.el**](https://github.com/bbatsov/prelude/blob/e0a86fcb02504b86a9f1fbe35bdbd837a47bd63e/core/prelude-packages.el#L62)
2. [**cask*'s **epl-upgrade** function is used to install/upgrade all packages on startup in **prelude-core.el**](https://github.com/bbatsov/prelude/blob/e0ca7c700389e70df457177ea75b0936dbe254e0/core/prelude-core.el#L155)
3. [**cask**'s **epl-upgrade** internally calls **package.el** with some update vs install auto-handling](https://cask.readthedocs.io/en/latest/guide/introduction.html)

Maybe something fails during the **dash** compile or that of previous packages in **prelude-packages** leaving the overall install. It's possible that there is a package depended upon by Emacs Prelude whose compile depends on external OS commands that are not enumerated correctly in the APT dependencies (remembering that APT is **not** the recommended or even documented approach to installing Prelude).

**Comment by AlainODea** [comment]
@zobbo you should install Emacs Prelude with the [Prelude automated installer](https://github.com/bbatsov/prelude#automated) and not with APT.

**Comment by Fuco1** [comment]
@AlainODea Thanks for the super detailed explanation! This makes a lot of sense :) 

The user in question also installed dash as an apt package which probably confuses the hell out of load-path :P

**Comment by AlainODea** [comment]
@Fuco1 my pleasure, friend. Feel free to tag me any further issues **dash** users encounter with Prelude. I might not be able to find the answer, but I'll be happy to help dig for it :)

**Comment by pigmej** [comment]
From: Matus Goljer__‎Sent: sobota, 18 lutego 2017 21:26To: bbatsov/preludeReply To: bbatsov/preludeCc: SubscribedSubject: Re: [bbatsov/prelude] Why is prelude not installed via package.el? (#1064)Closed #1064.

—You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/bbatsov/prelude","title":"bbatsov/prelude","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/bbatsov/prelude"}},"updates":{"snippets":[{"icon":"DESCRIPTION","message":"Closed #1064."}],"action":{"name":"View Issue","url":"https://github.com/bbatsov/prelude/issues/1064#event-968112651"}}}

**Comment by zobbo** [comment]
My dear friends - it is a pleasure to be mentioned here, but can I make one little thing clear. 

I have installed Prelude ALWAYS with the automated installer. I don't know of another way. For some reason, when I tried to install on mint it complained about dash. I don't know why - and because I was on a quest for doing other things, not prelude debugging, I "solved" my issue by using apt to get dash  - as the automated installed didn't cope with it in that instance. I have installed Prelude since multiple times as well on plain Ubuntu without needing to do this. I have not installed Mint again so don't know if the problem happens again there.

It might have been a glitch, it might have been a bug - I didn't know. I just reported it. I love Prelude and I love everyone who contributes and supports it - thank you :)

**Comment by AlainODea** [comment]
@zobbo no worries, friend. Reporting issues is the first step and I thank you for that. Using APT is a clever workaround and your reporting of it is useful since others finding the same issue will have a workaround a web search away :)

## tl;dr
Installing emacsen-common by itself may fix this issue on Linux Mint.

## The Long Version
I think default installed packages on Ubuntu and Linux Mint differ. As a result something needed to build or run the Emacs packages used by Prelude is present by default on Ubuntu and missing by default on Linux Mint. I suspect the answ

There is some potential insight in the differences between the emacs-nox (to keep the dependencies to the core) and emacs-el packages:
http://packages.ubuntu.com/xenial/emacs24-nox
http://packages.ubuntu.com/xenial/dash-el

Notably dash-el adds a novel dependency (one not depended upon by emacs-nox) on emacsen-common. emacsen-common has an enlightening description:

> This package contains code that is needed by all the (x)emacs packages. It will be automatically installed when needed.

The fact that this works on Ubuntu and fails on Linux Mint suggests that there are undeclared/implicit dependencies in Ubuntu's APT repo. That suggests that Emacs packages are working on Ubuntu because the undeclared dependencies happen to be present. Installing emacsen-common by itself may fix this issue on Linux Mint.

I entertained the possibility of diffing the default package lists for Ubuntu 16.10 and Linux Mint 18. It would be somewhere in there, but I suspect the diff will be massive.

**Comment by AlainODea** [comment]
@pigmej do you have an automation malfunctioning? I wasn't able to confirm that it leaked API keys, but it certainly looks like it has. If there are real API keys in your comment you should rotate them as soon as possible.
